### PR TITLE
[YGBWSVY-34] Fix user login issue

### PR DESCRIPTION
**Pull request by NightHunterSV**
https://fivejars.atlassian.net/browse/YGBWSVY-34

If a user, having non-VY roles, as well as virtual_y(_*) role(s) tries to log in using /user/login page -- he gets the error message: "You're not allowing to login with Virtual Y user role."
If he tries to log in using the Virtual Y log-in page -- he also not allowed to log in, with the message: "This user is not allowed to login from this form."
This PR fixes this issue -- so that the users with non-VY roles should be able to log in using the Drupal login form.

## Steps to test:

- [ ] Log in to the site as an admin
- [ ] Manage som user and assign non-VY and VY roles at the same time
- [ ] Log out or open the new session (use Incognito mode, etc.)
- [ ] Try to log in using VY login form -- you should get the error message.
- [ ] Try to log in using the `/user/login` page -- you should be able now.
- [ ] Log in as admin again
- [ ] Edit some user and assign it only VY roles
- [ ] Log out and try to log in using `/user/login` with the edited user
- [ ] Ensure you get the error message
- [ ] Try to log in using VY login form
- [ ] Ensure you've successfully logged in

### CODE REVIEW

**Review by hamrant — APPROVED** [review]
Code looks ok, @NightHunterSV  please add steps to check that the previous logic works fine

**Comment by NightHunterSV** [comment]
Done
Check, please

File: src/EditUserAccountAccessCheck.php [hunk 1]
```diff
@@ -26,10 +26,11 @@ class EditUserAccountAccessCheck implements AccessInterface {
    */
   public function access(AccountInterface $user, AccountInterface $account) {
     $roles = $account->getRoles();
-    foreach ($roles as $role) {
-      if (strpos($role, 'virtual_y_') !== FALSE || $role == 'virtual_y') {
-        return AccessResult::forbidden()->cachePerUser();
-      }
+    $vy_roles = array_filter($roles, function ($role) {
+      return strpos($role, 'virtual_y_') !== FALSE || $role == 'virtual_y';
+    });
```
Comment (anpolimus) [inline src/EditUserAccountAccessCheck.php:31]: why do we need to modify this place?

**Review by anpolimus — CHANGES_REQUESTED** [review]

File: src/EditUserAccountAccessCheck.php (hunk 1 above)
Comment (NightHunterSV) [reply src/EditUserAccountAccessCheck.php:31]: Was found that the user with admin/editor/whatever role can't edit user account if he has also `virtual_y` role

**Comment by anpolimus** [comment]
Marking it as Do not merge, since this fix is postponed for 1.3

File: openy_gated_content.module [hunk 2]
```diff
@@ -282,11 +282,12 @@ function openy_gated_content_user_login_validation(array &$form, FormStateInterf
...
-    foreach ($roles as $role) {
-      if (strpos($role, 'virtual_y_') !== FALSE || $role == 'virtual_y') {
-        \Drupal::messenger()->addError('You\'re not allowing to login with Virtual Y user role.');
-        throw new AccessDeniedHttpException();
-      }
+    $vy_roles = array_filter($roles, function ($role) {
+      return strpos($role, 'virtual_y_') !== FALSE || $role == 'virtual_y';
+    });
+    if (count(array_diff($roles, array_merge([AccountInterface::AUTHENTICATED_ROLE], $vy_roles))) === 0) {
+      \Drupal::messenger()->addError('You\'re not allowing to login with Virtual Y user role.');
```
Comment (froboy) [inline openy_gated_content.module:289]: I'd recommend cleaning up this message. Possibly:

> Virtual YMCA members are not permitted to change their password with this form. Please contact your branch for assistance or go back to the homepage to log in.

Comment (hamrant) [reply openy_gated_content.module:289]: @NightHunterSV and check please this comment - https://github.com/ymcatwincities/openy_gated_content/issues/65#issuecomment-763155763

**Comment by anpolimus** [comment]
@NightHunterSV I cant log in to the system as admin.
User login form shows 403 orrerror

**Review by hamrant — DISMISSED** [review]
Looks ok, let's test this

File: openy_gated_content.module [hunk 3]
```diff
@@ -282,11 +282,14 @@ function openy_gated_content_user_login_validation(array &$form, FormStateInterf
...
-        \Drupal::messenger()->addError('You\'re not allowing to login with Virtual Y user role.');
-        throw new AccessDeniedHttpException();
-      }
+    $vy_roles = array_filter($roles, function ($role) {
+      return strpos($role, 'virtual_y_') !== FALSE || $role === 'virtual_y';
+    });
+    if (count($vy_roles) > 0 &&
+      count(array_diff($roles, array_merge([AccountInterface::AUTHENTICATED_ROLE], $vy_roles))) === 0
+    ) {
+      \Drupal::messenger()->addError('You\'re not allowing to login with Virtual Y user role.');
```
Comment (anpolimus) [inline openy_gated_content.module:291]:
```suggestion
      \Drupal::messenger()->addError('Virtual YMCA members are not permitted to use this form. Please contact your branch for assistance or go back to the homepage to log in.');
```

**Comment by anpolimus** [comment]
Everything is ok. Ready to QA.

**Comment by AnastasiiaPys** [comment]
Tested, works as expected. thank you!

**Pull request merged**